### PR TITLE
Creates the Speedometer Widget UI and Funtionality 

### PR DESCRIPTION
**Pull request by Voldivh**
## Summary
This PR creates the Speedometer gauge with it's functionalities as the first widget necessary to completely address the second step of #1  . This widget is expected to be able to subscribe to a specific topic and select a field from that topic to show on the gauge UI. The following features are added to this widget:

- Show the value of the specified topic field as a speedometer.
- Modify the minimum and maximum value of the gauge
- Modify the units of the speedometer
## Run the widget
In order to test out this widget,  follow this next steps:

1. Compile the package and source the workspace:

```bash
cd <workspace-with-the-package>
clear && colcon build --packages-select rqt_gauges_2
source install/setup.bash
```
Here you can choose to run it on RQT or as a Standalone node:
 
### RQT
2. Run the RQT plugin UI:

```bash
rqt
```
3. Go to the following path on the menu tab: Plugins -> Visualization -> Speedometer
image

### Standalone Plugin
2. Run the standalone node:

```bash
ros2 run rqt_gauges_2 rqt_gauges_2
```
### Test it out
In order to test out, you can do the following:
- Play by changing the minimum limit, maximum limit and units of the gauge with the boxes at the bottom of the widget.
- Subscribe to topics and watch the gauge move:
1. Create a publisher other terminals and run:
```bash
ros2 topic pub /test_topic_2 geometry_msgs/msg/Twist "linear:
  x: 70.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"
```
2. Create another publisher on other terminal and run:
```bash
ros2 topic pub /test_topic std_msgs/msg/Int16 "data: 10"
```
3. Go to the plugin and start filling the text box on top of it until you reach the specific field you want for a specific topic (there is a topic autocompleter in order to avoid any name issues):
image
4. After selecting the topic and completing the required field, hit the Subscribe button to the right.
5. Take a look at the gauge move and play around  :grin:

#### PS
So sorry for the long PR :sob:

### CODE REVIEW

File: src/rqt_gauges_2/speedometer.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+def is_number_type(type):
+    number_types = [
+        'int8', 'uint8',
+        'int16', 'uint16',
+        'int32', 'uint32',
+        'int64', 'uint64',
+        'float', 'float32', 'float64',
+        'double', 'long double']
+    return type in number_types
```
Comment (francocipollone) [inline src/rqt_gauges_2/speedometer.py]: Are you using it?

File: src/rqt_gauges_2/speedometer.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class Speedometer(Plugin):
+
```
Comment (francocipollone) [inline src/rqt_gauges_2/speedometer.py:5]: A some docstring to all the classes and methods in this PR.

File: src/rqt_gauges_2/speedometer_widget.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def updateUnits(self, new_units):
+        self.speedometer_gauge.units = new_units
+        self.speedometer_gauge.update()
+
+    @pyqtSlot()
+    def updateSubscription(self):
+        if self.node.destroy_subscription(self.sub):
+            print("Previous subscription deleted")
+        else:
+            print("There was no previous subscription")
```
Comment (francocipollone) [inline src/rqt_gauges_2/speedometer_widget.py:118]: The other rqt plugins, what kinds of loggers are they using for these kinds of prints?

File: src/rqt_gauges_2/speedometer_gauge.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+        self.enable_fine_scaled_marker = True
+        self.enable_big_scaled_marker = True
+        self.enable_Needle_Polygon = True
+
+        self.needle_scale_factor = 0.8
+
+        self.gauge_color_outer_radius_factor = 1
+        self.gauge_color_inner_radius_factor = 0.9
+
+        # QWidget.setMouseTracking(self, True)
```
Comment (francocipollone) [inline src/rqt_gauges_2/speedometer_gauge.py]: Should it be removed?

**Review by francocipollone — COMMENTED** [review]
Great Job! I left some minor comments. I also tried it locally and I have some comments I will comment

**Comment by francocipollone** [comment]
### Discovery of topics
1. I opened the speedometer.
2. Then, I published some test data in some topics.
3. I can't select that topic in the speedometer plugin. It is like the discovery of topics happens only up to when the speedometer was opened. This isn't convenient. 

### Setting min and max value
It is weird to see the highest value not matching the max value that the user has entered:
image

**Comment by Voldivh** [comment]
@francocipollone 
> ### Discovery of topics
> 1. I opened the speedometer.
> 2. Then, I published some test data in some topics.
> 3. I can't select that topic in the speedometer plugin. It is like the discovery of topics happens only up to when the speedometer was opened. This isn't convenient.

I was aware of this behavior, I have a couple of options I wanted to discuss on how to solve this:
1. Create a refresh button on the widget that basically discovers all the topics at that moment. 
2. The object that I'm using to show the topics is a QLineEdit, we could use their [signals](https://doc.qt.io/qt-6/qlineedit.html#signals) to trigger the update function of the widget and that way we can have a discovery state "on demand". However, from the signals that object have, there isn't a clear choice since the signals that could be useful would impose a restriction to the user such as "Write/delete something in order to update the topics". I also think this isn't convenient.
3. Create a thread that updates the topics periodically. I think this solution can work but is a bit of an overkill.

Let me know what you think about this issue.


> ### Setting min and max value
> It is weird to see the highest value not matching the max value that the user has entered:

This actually seems weird to me, I tried exactly that and did't get any problems. Did you clicked Enter after writing down the number? I think I can modify it to better use the value change signal from the object, the question is just to appease my curiosity.

Edit: I just recalled this happens due to the segmentation of the whole circle into a semicircle. Due to the calculations made to readjust the circle and the amount of scale markers, the final value is not exactly at the end of the semicircle, unless the maximum number is a multiple of 90. I can take a look at this and see how to solve it. However, this doesn't mean that the big number in the middle of the widget won't be able to get to the specified maximum value.

File: src/rqt_gauges_2/speedometer_widget.py (hunk 3 above)
Comment (Voldivh) [reply src/rqt_gauges_2/speedometer_widget.py:118]: From checking [RQT_Plot](https://github.com/ros-visualization/rqt_plot/tree/foxy-devel/src/rqt_plot) I only noticed some prints the same way I'm doing it. However, I could get the node logger and log the same information.

File: src/rqt_gauges_2/speedometer_gauge.py (hunk 4 above)
Comment (Voldivh) [reply src/rqt_gauges_2/speedometer_gauge.py]: Yes it can. Thanks and done.

File: src/rqt_gauges_2/speedometer.py (hunk 1 above)
Comment (Voldivh) [reply src/rqt_gauges_2/speedometer.py]: Nope, just forgot to delete it. Thanks.

**Comment by francocipollone** [comment]
> 2. The object that I'm using to show the topics is a QLineEdit, we could use their [signals](https://doc.qt.io/qt-6/qlineedit.html#signals) to trigger the update function of the widget and that way we can have a discovery state "on demand". However, from the signals that object have, there isn't a clear choice since the signals that could be useful would impose a restriction to the user such as "Write/delete something in order to update the topics". I also think this isn't convenient.

I think that using a signal like FocusIn makes sense, in way that it only updates the list when you are working on that field. It sounds like the best option as we want it to be updated and we don't want to have a timer for that. (Maybe plotter gives us a hint.)

**Comment by francocipollone** [comment]
> Edit: I just recalled this happens due to the segmentation of the whole circle into a semicircle. Due to the calculations made to readjust the circle and the amount of scale markers, the final value is not exactly at the end of the semicircle, unless the maximum number is a multiple of 90. I can take a look at this and see how to solve it. However, this doesn't mean that the big number in the middle of the widget won't be able to get to the specified maximum value.

Makes sense. It is ok then. I will fill up a ticket later on just to make it a known issue. There is no necessity on tackling this, it is just an enhancement.

**Comment by francocipollone** [comment]
Let me know when it is ready for another pass. (In case you are missing some cycles to work on this, let me know and I will try to continue)

**Review by francocipollone — APPROVED** [review]
Great job! Works like a charm

**Pull request merged**